contracts/gov/proposals/TokenSaleProposal.sol

```diff
@@ -0,0 +1,381 @@
+            "TSP: purchase tokens are not provided"
+        );
+        require(
+            tierView.purchaseTokenAddresses.length == tierView.exchangeRates.length,
+            "TSP: tokens and rates lens mismatch"
+        );
+
+        for (uint256 i = 0; i < tierView.purchaseTokenAddresses.length; i++) {
+            require(tierView.exchangeRates[i] != 0, "TSP: rate cannot be zero");
+
+            require(
+                tierView.purchaseTokenAddresses[i] != address(0),
+                "TSP: purchase token cannot be zero"
+            );
+
+            require(
+                tierInfo.rates[tierView.purchaseTokenAddresses[i]] == 0,
+                "TSP: purchase tokens are duplicated"
+            );
+
+            tierInfo.rates[tierView.purchaseTokenAddresses[i]] = tierView.exchangeRates[i];
+        }
+    }
+
+    function _addToWhitelist(
+        WhitelistingRequest calldata request
+    ) internal ifTierExists(request.tierId) {
+        for (uint256 i = 0; i < request.users.length; i++) {
+            _mint(request.users[i], request.tierId, 1, "");
+        }
+    }
+
+    function _offTier(uint256 tierId) internal ifTierExists(tierId) {
+        TierInfo storage tierInfo = _tiers[tierId].tierInfo;
+
+        require(!tierInfo.isOff, "TSP: tier is already off");
+
+        tierInfo.isOff = true;
+    }
+
+    function _getVestingWithdrawAmount(
+        uint256 tierId
+    ) internal view ifTierExists(tierId) returns (uint256) {
+        Tier storage tier = _tiers[tierId];
+
+        TierView memory tierView = tier.tierView;
+        TierInfo storage tierInfo = tier.tierInfo;
+
+        if (tierInfo.customers[msg.sender].purchaseTime == 0) {
+            return 0;
+        }
+
+        VestingSettings memory vestingSettings = tierView.vestingSettings;
+        Purchase memory purchase = tierInfo.customers[msg.sender];
+
+        uint256 startTime = purchase.purchaseTime + vestingSettings.cliffPeriod;
+
+        if (startTime > block.timestamp) {
+            return 0;
+        }
+
+        uint256 stepsCount = vestingSettings.vestingDuration / vestingSettings.unlockStep;
+        uint256 tokensPerStep = purchase.vestingAmount / stepsCount;
+
+        uint256 vestingWithdrawAmount = (block.timestamp - purchase.latestVestingWithdraw).ratio(
+            tokensPerStep,
+            vestingSettings.unlockStep
+        );
+
+        return vestingWithdrawAmount;
+    }
+
+    function _getRecoverAmount(
+        uint256 tierId
+    ) internal view ifTierExists(tierId) returns (uint256) {
+        TierView memory tierView = _tiers[tierId].tierView;
+
+        if (block.timestamp <= tierView.saleEndTime) {
+            return 0;
+        }
+
+        uint256 balanceLeft = tierView.totalTokenProvided - _tiers[tierId].tierInfo.totalSold;
+
+        return
+            ERC20(tierView.saleTokenAddress)
+                .balanceOf(address(this))
+                .to18(ERC20(tierView.saleTokenAddress).decimals())
```

Comment:
Let's make the contract work with the NON-NORMALIZED decimals cause it might be hard to create normalized proposals

Reply:
done